Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm

```diff
@@ -193,38 +209,47 @@ static float toNSURLSessionTaskPriority(WebCore::ResourceLoadPriority priority)
+    m_task.get()._cookieTransformCallback = makeBlockPtr([requestURL = crossThreadCopy(request.url()), firstPartyURL = crossThreadCopy(firstPartyURL), firstPartyHostCNAME = crossThreadCopy(firstPartyHostCNAME), firstPartyAddress = crossThreadCopy(firstPartyAddress), thirdPartyCNAMEDomainForTesting = crossThreadCopy(networkSession()->thirdPartyCNAMEDomainForTesting()), ageCapForCNAMECloakedCookies = crossThreadCopy(m_ageCapForCNAMECloakedCookies), weakTask = WeakObjCPtr<NSURLSessionDataTask>(m_task.get()), debugLoggingEnabled = networkSession()->networkStorageSession()->resourceLoadStatisticsDebugLoggingEnabled()] (NSArray<NSHTTPCookie*> *cookiesSetInResponse) -> NSArray<NSHTTPCookie*> * {
         auto task = weakTask.get();
         if (!task || ![cookiesSetInResponse count])
             return cookiesSetInResponse;
 
         auto cnameDomain = lastCNAMEDomain([task _resolvedCNAMEChain]);
+        if (cnameDomain.isEmpty() && thirdPartyCNAMEDomainForTesting)
+            cnameDomain = *thirdPartyCNAMEDomainForTesting;
+
         if (cnameDomain.isEmpty()) {
-            if (!thirdPartyCNAMEDomainForTesting)
+            if (!firstPartyAddress)
                 return cookiesSetInResponse;
-            cnameDomain = *thirdPartyCNAMEDomainForTesting;
+
+            auto remoteAddress = WebCore::IPAddress::fromString(lastRemoteIPAddress(task.get()));
+            if (!remoteAddress)
+                return cookiesSetInResponse;
+
+            if (shouldCapCookieExpiry(*remoteAddress, *firstPartyAddress)) {
+                cookiesSetInResponse = cookiesByCappingExpiry(cookiesSetInResponse, ageCapForCNAMECloakedCookies);
+                if (debugLoggingEnabled) {
+                    for (NSHTTPCookie *cookie in cookiesSetInResponse)
+                        RELEASE_LOG_INFO(ITPDebug, "Capped the expiry of third-party IP address cookie named %{public}@.", cookie.name);
+                }
+            }
+
+            return cookiesSetInResponse;
         }
 
         // CNAME cloaking is a first-party sub resource that resolves
         // through a CNAME that differs from the first-party domain and
         // also differs from the top frame host's CNAME, if one exists.
         if (!cnameDomain.matches(firstPartyURL) && (!firstPartyHostCNAME || cnameDomain != *firstPartyHostCNAME)) {
-
-            NSUInteger count = [cookiesSetInResponse count];
-            // Don't use RetainPtr here. This array has to be retained and
```

Comment:
I remember adding this comment (and changing the code accordingly) because of an issue in CFNetwork's code. Have you made sure we're not regressing that now? I think it was for the zero cookie case. Please retain the comment if it is still relevant because it took us a long time to debug.

Reply:
Sounds good — I put this comment back.

I believe what the comment is _probably_ referring to is how CFNetwork doesn't immediately retain the array we hand it back in the cookie transform block; this means that if we use a `RetainPtr` here, the array will be destroyed after the function call, before CFNetwork actually uses it, which would cause a UAF.